src/hotspot/share/prims/agentList.cpp

```diff
@@ -0,0 +1,652 @@
+void AgentList::convert_xrun_agents() {
+  Iterator it = xrun_agents();
+  while (it.has_next()) {
+    Agent* const agent = it.next();
+    assert(agent->_is_xrun, "invariant");
+    OnLoadEntry_t on_load_entry = lookup_JVM_OnLoad_entry_point(agent);
+    // If there is an JVM_OnLoad function it will get called later,
+    // otherwise see if there is an Agent_OnLoad.
+    if (on_load_entry == nullptr) {
+      on_load_entry = lookup_Agent_OnLoad_entry_point(agent);
+      if (on_load_entry == nullptr) {
+        vm_exit_during_initialization("Could not find JVM_OnLoad or Agent_OnLoad function in the library", agent->name());
+      }
+      agent->_is_xrun = false; // converted
+    }
+  }
+}
+
+// Invokes Agent_OnLoad for -agentlib:.. -agentpath:  and converted -Xrun agents.
+// Called very early -- before JavaThreads exist
+void AgentList::load_agents() {
+  assert(JvmtiEnvBase::get_phase() == JVMTI_PHASE_PRIMORDIAL, "invalid init sequence");
+  extern struct JavaVM_ main_vm;
+
+  // Convert -Xrun to -agentlib: if there is no JVM_OnLoad
+  convert_xrun_agents();
+
+  JvmtiExport::enter_onload_phase();
+
+  Iterator it = agents();
+  while (it.has_next()) {
+    Agent* agent = it.next();
+    if (Arguments::is_dumping_archive()) {
+      check_cds_dump(agent);
+    }
+
+    OnLoadEntry_t on_load_entry = lookup_Agent_OnLoad_entry_point(agent);
+
+    if (on_load_entry == nullptr) {
+      vm_exit_during_initialization("Could not find Agent_OnLoad function in the agent library", agent->name());
+    }
+    // Invoke the Agent_OnLoad function
+    if ((*on_load_entry)(&main_vm, const_cast<char*>(agent->options()), nullptr) != JNI_OK) {
+      vm_exit_during_initialization("agent library failed Agent_OnLoad", agent->name());
+    }
+
+    // Convert the instrument lib to the actual JPLIS / javaagent it represents.
+    if (agent->is_instrument_lib()) {
+      convert_to_jplis(agent);
+    }
+  }
+
+  JvmtiExport::enter_primordial_phase();
+}
+
+extern "C" {
+  typedef void (JNICALL* Agent_OnUnload_t)(JavaVM*);
+}
+
+// Called after the VM is initialized for -Xrun agents which have not been converted to JVMTI agents
+void AgentList::invoke_JVM_OnLoad() {
+  extern struct JavaVM_ main_vm;
+  Iterator it = xrun_agents();
+  while (it.has_next()) {
+    Agent* agent = it.next();
+    assert(agent->_is_xrun, "invariant");
+    OnLoadEntry_t on_load_entry = lookup_JVM_OnLoad_entry_point(agent);
+    if (on_load_entry == nullptr) {
+      vm_exit_during_initialization("Could not find JVM_OnLoad function in -Xrun library", agent->name());
+    }
+    // Invoke the JVM_OnLoad function
+    JavaThread* thread = JavaThread::current();
+    ThreadToNativeFromVM ttn(thread);
+    HandleMark hm(thread);
+    const jint err = (*on_load_entry)(&main_vm, const_cast<char*>(agent->options()), NULL);
+    if (err != JNI_OK) {
+      vm_exit_during_initialization("-Xrun library failed to init", agent->name());
```

Comment:
Do you need to be back in `_thread_in_vm` before exiting?

Reply:
Hmm. This was ported as is. I will double-check.